service-workers/service-worker/waiting.https.html

```diff
@@ -5,36 +5,40 @@
 <script src="resources/test-helpers.sub.js"></script>
 <body>
 <script>
-promise_test(t => {
-    var script = 'resources/empty-worker.js';
-    var scope = 'resources/blank.html';
-    var frame;
-    var registration;
 
-    return service_worker_unregister(t, scope)
-      .then(() => { return with_iframe(scope); })
-      .then(f => {
-          frame = f;
-          return navigator.serviceWorker.register(script, {scope: scope});
-        })
-      .then(r => {
-          registration = r;
-          add_result_callback(() => { registration.unregister(); });
-          return wait_for_state(t, r.installing, 'installed');
-        })
-      .then(() => {
-          const controller =
-            frame.contentWindow.navigator.serviceWorker.controller;
-          assert_equals(controller, null, '.controller');
+const SCRIPT = 'resources/empty-worker.js';
+const SCOPE = 'resources/blank.html';
 
-          assert_equals(registration.active, null, '.active');
-          assert_equals(registration.waiting.state, 'installed', '.waiting');
-          assert_equals(registration.installing, null, '.installing');
-        })
-      .then(() => {
-          frame.remove();
-          return registration.unregister();
-        });
-  }, 'waiting is set after installation');
+promise_test(async t => {
+  await service_worker_unregister(t, SCOPE);
+  const frame = await with_iframe(SCOPE);
+  const registration =
+      await navigator.serviceWorker.register(SCRIPT, {scope: SCOPE});
+  await wait_for_state(t, registration.installing, 'installed');
+  const controller = frame.contentWindow.navigator.serviceWorker.controller;
+
+  assert_equals(controller, null, 'controller');
+  assert_equals(registration.active, null, 'registration.active');
+  assert_equals(registration.waiting.state, 'installed',
+                'registration.waiting');
+  assert_equals(registration.installing, null, 'registration.installing');
+  t.add_cleanup(async() => {
+    frame.remove();
+    await service_worker_unregister_and_done(t, SCOPE);
```

Comment:
could we make the cleanup function like this:

t.add_cleanup(async() => {
  if (frame)
    frame.remove();
  if (registration)
    await registration.unregister();
});

It'd be better if it's moved to the top of the function so in case an assert fails, the cleanup is still run. I'm not sure if it would have to go after registration and frame are defined.